### PR TITLE
Room permissions

### PR DESCRIPTION
**Pull request by jbrunton**

### CODE REVIEW

File: services/api/src/app/auth/permissions/roles.spec.ts [hunk 1]
```diff
@@ -19,7 +20,12 @@ describe('defineRolesForUser', () => {
   });
 
   it('grants read and write permissions for joined rooms', () => {
-    const room = subject('Room', RoomFactory.build());
+    const room = subject(
+      'Room',
+      RoomFactory.build({
+        contentPolicy: ContentPolicy.Private,
+      }),
+    );
```
Comment (jbrunton) [inline services/api/src/app/auth/permissions/roles.spec.ts]: Unnecessary subject() call

**Comment by jbrunton-ci-minion** [comment]
#### :tropical_drink: `up` on auth0-test/room-permissions

   <details>
   <summary>Pulumi report</summary>

   ```
   Updating (room-permissions)

View Live: https://app.pulumi.com/jbrunton/auth0-test/room-permissions/updates/4

@ Updating........
   pulumi:pulumi:Stack auth0-test-room-permissions  warning: provider config warning: With the retirement of EC2-Classic the skip_get_ec2_platforms attribute has been deprecated and will be removed in a future version.

@ Updating....
   pulumi:pulumi:Stack auth0-test-room-permissions running warning: provider config warning: With the retirement of EC2-Classic the skip_get_ec2_platforms attribute has been deprecated and will be removed in a future version.
@ Updating..............
   pulumi:pulumi:Stack auth0-test-room-permissions running app config: {
   pulumi:providers:aws aws  
   pulumi:pulumi:Stack auth0-test-room-permissions running read pulumi:pulumi:StackReference jbrunton/jbrunton-aws.com-infra/prod
   pulumiservice:index:StackTag stack-tags  
   pulumi:pulumi:Stack auth0-test-room-permissions running read pulumi:pulumi:StackReference jbrunton/jbrunton-aws.com-infra/prod
@ Updating....
   aws:ecs:Cluster auth0-test-room-permissi  
   aws:cloudwatch:LogGroup /ecs/auth0-test-room-permissions-api  
   aws:lb:TargetGroup auth0-test-room-permissi  
   aws:dynamodb:Table auth0-test-room-permissions-api  
   aws:iam:Role auth0-test-room-permissions-api-exec-role  
~  aws:s3:Bucket auth0-test-room-permissions-bucket updating (0s) [diff: -serverSideEncryptionConfiguration]
   aws:iam:Role auth0-test-room-permissions-api-role  
   aws:ec2:SecurityGroup auth0-test-room-permissions-api-sg  
   aws:lb:ListenerRule auth0-test-room-permissions-api  
   aws:iam:RolePolicyAttachment auth0-test-room-permissions-api-exec-role-attachment  
   aws:iam:RolePolicy auth0-test-room-permissions-api-get-params  
   aws:iam:RolePolicy auth0-test-room-permissions-api-dynamo-db  
++ aws:ecs:TaskDefinition auth0-test-room-permissions creating replacement (0s) [diff: ~containerDefinitions]
   aws:iam:RolePolicyAttachment auth0-test-room-permissions-api-role-attachment  
   aws:route53:Record auth0-test-room-permissions  
@ Updating....
++ aws:ecs:TaskDefinition auth0-test-room-permissions created replacement (0.53s) [diff: ~containerDefinitions]
+- aws:ecs:TaskDefinition auth0-test-room-permissions replacing (0s) [diff: ~containerDefinitions]
+- aws:ecs:TaskDefinition auth0-test-room-permissions replaced (0.00s) [diff: ~containerDefinitions]
~  aws:ecs:Service auth0-test-room-permissions updating (0s) [diff: ~taskDefinition]
~  aws:s3:Bucket auth0-test-room-permissions-bucket updated (1s) [diff: -serverSideEncryptionConfiguration]
   aws:cloudfront:Distribution auth0-test-room-permissions-cdn  
~  aws:ecs:Service auth0-test-room-permissions updated (0.55s) [diff: ~taskDefinition]
@ Updating....
   aws:route53:Record auth0-test-room-permissions.dev.jbrunton-aws.com  
   synced-folder:index:S3BucketFolder auth0-test-room-permissions-folder  
@ Updating....
   aws:s3:BucketObject vite.svg  
   aws:s3:BucketObject assets/index-22545a0c.css  
   aws:s3:BucketObject index.html  
   aws:s3:BucketObject assets/index-5e3dc46d.js  
-- aws:ecs:TaskDefinition auth0-test-room-permissions deleting original (0s) [diff: ~containerDefinitions]
-- aws:ecs:TaskDefinition auth0-test-room-permissions deleted original (0.21s) [diff: ~containerDefinitions]
   pulumi:pulumi:Stack auth0-test-room-permissions  1 warning; 1 message

Diagnostics:
 pulumi:pulumi:Stack (auth0-test-room-permissions):
   warning: provider config warning: With the retirement of EC2-Classic the skip_get_ec2_platforms attribute has been deprecated and will be removed in a future version.
   app config: {
    "stackName": "room-permissions",
    "client": {
     "name": "client"
    },
    "appName": "auth0-test-room-permissions",
    "environment": "development",
    "services": [
     {
      "name": "api",
      "repository": "auth0-test-api",
      "tag": "HEAD-14263c3"
     }
    ],
    "protect": false
   }

Outputs:
   publicApiUrl: "https://auth0-test-api-room-permissions.dev.jbrunton-aws.com"
   publicUrl   : "https://auth0-test-room-permissions.dev.jbrunton-aws.com"

Resources:
   ~ 2 updated
   +-1 replaced
   3 changes. 23 unchanged

Duration: 16s

   ```
   
   </details>

File: services/api/src/app/auth/auth.service.ts [hunk 2]
```diff
@@ -1,20 +1,47 @@
...
+      actions.map(async (action) => {
+        const can = await this.can({ user, subject, action });
+        if (can) {
+          return action;
+        } else {
+          return undefined;
+        }
+      }),
+    );
+    console.log('authorizedRoles:', { user, subject, roles });
```
Comment (jbrunton) [inline services/api/src/app/auth/auth.service.ts]: Remove this log

File: services/api/src/app/auth/auth.service.ts [hunk 3]
```diff
@@ -1,20 +1,47 @@
...
+  }
+
+  async authorizedRoles({
+    user,
+    subject,
+  }: Omit<AuthorizeParams, 'message' | 'action'>): Promise<Role[]> {
+    const actions: Role[] = [Role.Read, Role.Write, Role.Manage];
+    const roles = await Promise.all(
+      actions.map(async (action) => {
+        const can = await this.can({ user, subject, action });
```
Comment (jbrunton) [inline services/api/src/app/auth/auth.service.ts]: Call this `hasRole`

File: services/api/src/app/auth/auth.service.ts [hunk 4]
```diff
@@ -1,20 +1,47 @@
...
 @Injectable()
 export class CaslAuthService implements AuthService {
   constructor(private readonly membershipsRepo: MembershipsRepository) {}
 
-  async authorize({ user, room, action, message }: AuthorizeParams) {
-    const memberships = await this.membershipsRepo.getMemberships(user.id);
-    const ability = defineRolesForUser(user, memberships);
-    if (!ability.can(action, subject('Room', room))) {
+  async authorize({ user, subject, action, message }: AuthorizeParams) {
+    if (!(await this.can({ user, action, subject }))) {
```
Comment (jbrunton) [inline services/api/src/app/auth/auth.service.ts]: Refactor

**Pull request merged**